include/ck/tensor_operation/gpu/grid/gridwise_add_layernorm_welford_variance.hpp

```diff
@@ -0,0 +1,424 @@
+
+        // Copy x from Cache
+        // one pass: fwd, second pass: bwd
+        constexpr auto thread_copy_fwd_step_k = make_multi_index(SweepOnce ? 0 : K_BlockTileSize);
+        constexpr auto thread_copy_bwd_step_k = make_multi_index(SweepOnce ? 0 : -K_BlockTileSize);
+
+        constexpr auto thread_copy_fwd_step_m_k =
+            make_multi_index(0, SweepOnce ? 0 : K_BlockTileSize);
+        constexpr auto thread_copy_bwd_step_m_k =
+            make_multi_index(0, SweepOnce ? 0 : -K_BlockTileSize);
+
+        const auto a_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_a_global, a_grid_desc_m_k.GetElementSpaceSize());
+
+        const auto b_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_b_global, b_grid_desc_m_k.GetElementSpaceSize());
+
+        const auto gamma_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_gamma_global, gamma_grid_desc_k.GetElementSpaceSize());
+
+        const auto beta_global_val_buf = make_dynamic_buffer<AddressSpaceEnum::Global>(
+            p_beta_global, beta_grid_desc_k.GetElementSpaceSize());
+
+        auto threadwise_welford       = ThreadwiseWelford();
+        threadwise_welford.max_count_ = GetKPerThread(c_grid_desc_m_k, thread_k_cluster_id);
+
+        static_for<0, MThreadSliceSize, 1>{}([&](auto I) {
+            mean_thread_buf(I) = type_convert<AccDataType>(0.0f);
+            var_thread_buf(I)  = type_convert<AccDataType>(0.0f);
+        });
+
+        for(index_t reducedTiles = 0; reducedTiles < num_k_block_tile_iteration; ++reducedTiles)
+        {
+
+            threadwise_a_load.Run(a_grid_desc_m_k,
+                                  a_global_val_buf,
+                                  thread_buffer_desc_m_k,
+                                  make_tuple(I0, I0),
+                                  a_thread_buf);
+
+            threadwise_b_load.Run(b_grid_desc_m_k,
+                                  b_global_val_buf,
+                                  thread_buffer_desc_m_k,
+                                  make_tuple(I0, I0),
+                                  b_thread_buf);
+
+            static_for<0, MThreadSliceSize, 1>{}([&](auto iM) {
+                static_for<0, KThreadSliceSize, 1>{}([&](auto iK) {
+                    constexpr auto offset_m_k =
+                        thread_buffer_desc_m_k.CalculateOffset(make_tuple(iM, iK));
+                    elementwise_op(c_thread_buf(Number<offset_m_k>{}),
+                                   a_thread_buf(Number<offset_m_k>{}),
+                                   b_thread_buf(Number<offset_m_k>{}));
+                    // c_thread_buf(Number<offset_m_k>{}) = a_thread_buf(Number<offset_m_k>{}) +
+                    // b_thread_buf(Number<offset_m_k>{});
+                });
+            });
+            threadwise_welford.Run(c_thread_buf, mean_thread_buf, var_thread_buf);
+
+            threadwise_c_store.Run(thread_buffer_desc_m_k,

```

Comment:
I am curious do we need to output the intermediate result for the fuse kernel?

Reply:
Yes, I think we need. In the current method, there are two loops on K dimension. The first loop do adding and use the adding result calculate means and vars. And the second loop will use means, vars and the final result calculate the final result. If K is larger than KThreadClusterSize*KThreadSliceSize, the current vgpr will be flashed and we need to reread the adding results. So I store the intermediate result into global memory for rereading in the second loop.

Reply:
Acutally, I think it's no need to store it when K is smaller than KThreadClusterSize*KThreadSliceSize. But as I try to reuse the adding results in vgpr when K is smaller, the numerical error of final results will be greater. I think the reason may be that when we declare StaticBuffer in vgpr to calculate the final result, the datatype we use is AccDatatype(FP32), but for the elementwise_add, the output result datatype is FP16. There may exist some problem in datatype transfer. But I haven't made some experiments to prove it. So I just store it in global memory and reread it into vgpr to FP32.